sdk/monitor/OpenTelemetry.Exporter.AzureMonitor/src/ConnectionString/ConnectionStringParser.cs

```diff
@@ -0,0 +1,124 @@
+                throw new Exception("Connection String Error: " + ex.Message, ex);
+            }
+        }
+
+        internal static string GetInstrumentationKey(this Azure.Core.ConnectionString connectionString) => connectionString.GetRequired(Constants.InstrumentationKeyKey);
+
+        /// <summary>
+        /// Will evaluate connection string and return the requested endpoint.
+        /// </summary>
+        /// <remarks>
+        /// Parsing the connection string MUST follow these rules:
+        ///     1. check for explicit endpoint (location is ignored)
+        ///     2. check for endpoint suffix (location is optional)
+        ///     3. use default endpoint (location is ignored)
+        /// This behavior is required by the Connection String Specification.
+        /// </remarks>
+        internal static string GetIngestionEndpoint(this Azure.Core.ConnectionString connectionString)
+        {
+            // Passing the user input values through the Uri constructor will verify that we've built a valid endpoint.
+            Uri uri;
+
+            if (connectionString.TryGetNonRequiredValue(Constants.IngestionExplicitEndpointKey, out string explicitEndpoint))
+            {
+                if (!Uri.TryCreate(explicitEndpoint, UriKind.Absolute, out uri))
+                {
+                    throw new ArgumentException($"The value for {Constants.IngestionExplicitEndpointKey} is invalid. '{explicitEndpoint}'");
+                }
+            }
+            else if (connectionString.TryGetNonRequiredValue(Constants.EndpointSuffixKey, out string endpointSuffix))
+            {
+                var location = connectionString.GetNonRequired(Constants.LocationKey);
+                if (!TryBuildUri(prefix: Constants.IngestionPrefix, suffix: endpointSuffix, location: location, uri: out uri))
+                {
+                    throw new ArgumentException($"The value for {Constants.EndpointSuffixKey} is invalid. '{endpointSuffix}'");
+                }
+            }
+            else
+            {
+                return Constants.DefaultIngestionEndpoint;
+            }
+
+            return uri.AbsoluteUri;
+        }
+
+        /// <summary>
+        /// Construct a Uri from the possible parts.
+        /// Format: "location.prefix.suffix".
+        /// Example: "https://westus2.dc.applicationinsights.azure.cn/".
+        /// </summary>
+        /// <remarks>
+        /// Will also attempt to sanitize user input. Won't fail if the user typo-ed an extra period.
+        /// </remarks>
+        internal static bool TryBuildUri(string prefix, string suffix, out Uri uri, string location = null)
+        {
+            // Location and Suffix are user input fields and need to be sanitized (extra spaces or periods).
+            char[] trimPeriod = new char[] { '.' };
+
+            if (location != null)
+            {
+                location = location.Trim().TrimEnd(trimPeriod);
```

Comment:
If we use `location?` we could reduce nesting by one level.

Reply:
yes, but we're working with the location variable twice; once for the Trim and once for the special chars check.
I think the current `if` statement is the better approach because it performs the null check only once.

Please comment if I misunderstood your suggestion.